### PR TITLE
runtimetest: add rootfs propagation test

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

### CODE REVIEW

File: cmd/runtimetest/main.go [hunk 1]
```diff
@@ -778,6 +843,14 @@ func run(context *cli.Context) error {
 			description: "read only paths",
 		},
 		{
+			test:        validateRootfsPropagation,
+			description: "rootfs propagation",
+		},
+		{
+			test:        validateRlimits,
```
Comment (wking) [inline cmd/runtimetest/main.go]: Duplicate entry?

```
$ git grep -n validateRlimits origin/pr/511
origin/pr/511:cmd/runtimetest/main.go:271:func validateRlimits(spec *rspec.Spec) error {
origin/pr/511:cmd/runtimetest/main.go:803:                      test:        validateRlimits,
origin/pr/511:cmd/runtimetest/main.go:850:                      test:        validateRlimits,
```

File: cmd/runtimetest/main.go [hunk 2]
```diff
@@ -342,6 +344,69 @@ func validateRootFS(spec *rspec.Spec) error {
...
+		return nil
+	}
+
+	targetDir, err := ioutil.TempDir("/", "target")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(targetDir)
+
+	switch spec.Linux.RootfsPropagation {
```
Comment (wking) [inline cmd/runtimetest/main.go]: I think we probably want a `default` case that logs an “unrecognized linux.rootfsPropagation: …” warning.

File: cmd/runtimetest/main.go (hunk 1 above)
Comment (Mashimiao) [reply cmd/runtimetest/main.go]: Oh, this is mistake, updated. Thanks.

File: cmd/runtimetest/main.go [hunk 3]
```diff
@@ -342,6 +344,71 @@ func validateRootFS(spec *rspec.Spec) error {
 	return nil
 }
 
+func validateRootfsPropagation(spec *rspec.Spec) error {
+	if spec.Linux == nil {
+		return nil
+	}
+
+	targetDir, err := ioutil.TempDir("/", "target")
```
Comment (q384566678) [inline cmd/runtimetest/main.go]: Can you change the create path? Perhaps `ioutil.TempDir("", "target")`? If you do not change, `runtimetest` must use `sudo`.

**Comment by Mashimiao** [comment]
On 11/04/2017 12:36 PM, W. Trevor King wrote:
> These tests don't seem to distinguish between |slave| and |private|.

Yes, I have to admit that we can't in this PR.
Because we can't get the mount source of rootfs.
So, we can just judge propagation by binding rootfs.
Unfortunately, we get the same behaviour with slave and private rootfs after binding them.

Maybe we can find a way and fix this in the future.

**Comment by liangchenye** [comment]
According to https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt,
This PR does tests 'share and unbind' mode. 
I think we can remove `TestValidateRootfsPropagationPrivate` and `TestValidateRootfsPropagationSlave`.

**Comment by liangchenye** [comment]
The 'share' and 'unbind' looks good to me, just need to rebase.

**Comment by liangchenye** [comment]
LGTM

**Comment by q384566678** [comment]
LGTM

**Pull request merged**